File: Problems/567_Permutation_in_String.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1)>len(s2):
            return False
        window = len(s1)
        alpha_table_target = [0]*26
        for c in s1:
            alpha_table_target[ord(c)-ord('a')] += 1

        alpha_table_current = [0]*26
        for i in range(window):
            alpha_table_current[ord(s2[i])-ord('a')] += 1

        if alpha_table_current == alpha_table_target:
            return True
        for i in range(window, len(s2)):
            alpha_table_current[ord(s2[i])-ord('a')]+=1
            alpha_table_current[ord(s2[i-window])-ord('a')]-=1
            if alpha_table_current == alpha_table_target:
                return True
        return False

if __name__ == '__main__':
    sol = Solution()
    s1 = "a"
    s2 = "ab"
    # s2 = "eidbaooo"
    print(sol.checkInclusion(s1, s2))

File: Problems/567/567_Permutation_in_String.cpp
#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.length()>s2.length()) return false;
        vector<int> alpha_table_1(26, 0);
        vector<int> alpha_table_2(26, 0);
        for(auto c: s1) alpha_table_1[c-'a']++;
        for(int i=0; i<s1.size(); i++) alpha_table_2[s2[i]-'a']++;
        if(alpha_table_1==alpha_table_2) return true;
        for(int i=s1.length(); i<s2.length(); i++){
            alpha_table_2[s2[i]-'a']++;
            alpha_table_2[s2[i-s1.length()]-'a']--;
            if(alpha_table_1==alpha_table_2) return true;
        }
        return false;
    }
};